kep/640-error-pods-rescheduling/README.md

```diff
@@ -0,0 +1,166 @@
+Therefore, we are motivated to develop rescheduling plugins based on the proposed scheduler framework. These plugins can reschedule pods that encounter errors to different nodes, excluding the nodes where they were previously scheduled.
+
+### Goals
+
+* Reschedule already scheduled pods (encounter some errors) to other nodes automatically.
+* Implement pod rescheduling plugins in PreFilter extension point to prefilter nodes that the pod has already scheduled.
+* Pod rescheduling plugin in this proposal has no conflict with default scheduler's plugins.
+
+### Non-Goals
+
+* Reschedule pods controlled by daemonset or other workload resources which don't exist in default k8s controller.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+A pod submitted to the Kubernetes cluster initially operates smoothly, but over time, as the memory on the node hosting the pod becomes insufficient and the pod's memory requirements grow, it eventually gets CrashLoopBackOff due to out-of-memory (OOM) issues. 
+Consequently, we have to manually relocate the pod to other nodes. But this pod may schedule to the same node with the default scheduler.
+
+#### Story 2
+
+After a pod is submitted to the Kubernetes cluster and successfully scheduled, it may encounter startup issues due to mismatches in hardware resources on the node, 
+such as NUMA or device-related information. In such cases, we need to manually reschedule the pod to a different node. But this pod may schedule to the same node with the default scheduler.
+
+## Design Details
+
+Our scheme is to design two components as outlined in the diagram and described below: Rescheduler-controller and Podrescheduling. 
+Users can configure the `scheduling-retries` (maximum rescheduling retry count) in the metadata annotations of workload resources, such as Deployment and StatefulSet. 
+Rescheduler-controller is responsible for monitoring pods with abnormal states in the cluster. For each abnormal pod, 
+the rescheduler-controller checks the current scheduling count (defined as `currentReschedulingTimes`) recorded in the metadata annotations of its associated workload resource. 
+If the current scheduling count does not exceed the user-configured `scheduling-retries` (maximum rescheduling retry count), 
+it increments the current scheduling count by 1 and records the current node where the pod is located in the scheduled nodes list (eg: defined as `deployScheduledHosts`). 
+Then, rescheduler-controller deletes the abnormal pod, and this pod's associated workload resource will create a new pod automatically. 
+Rescheduler-controller registers a MutateWebHook in the kube-apiserver, this MutateWebHook will patch scheduled nodes list info to this newly created pod's annotations (defined as `sigs.k8s.io/scheduled-hosts`) when this newly created pod comes to kube-apiserver admission point.  
```

Comment:
I think I get why you want to do like this, but deleting a pod and recreating it overlaps with descheduler and other components. We need to clarify why descheduler is not good enough and/or how can be extended. Needed to do very similarly in https://docs.google.com/document/d/1-wJhiNy84w7tzFdo9HqwTu5DrVSuXFLGTUv8FBiRAAc/edit#heading=h.5mgn2182hlz6 - Please check "discarded alternatives" in the end.

Reply:
Ok, i will read the contents of this link carefully.

Reply:
The descheduler project pays attentio to **already running pods**, but our proposal focus on **pods running normally at first but fails after some time**.

Reply:
That would make sense to me. If this is the case also for your proposal, please make sure to write it down very clearly that in your usecase you cannot use descheduler for the reasons above.

Reply:
Ok, i will clarify it clearly in my KEP.